File: backend/internal/service/pokeapi_service.go
package service

import (
	"context"
	"encoding/json"
	"fmt"
	"math/rand"
	"net/http"
	"sort"
	"strings"
	"sync"

	"github.com/kenyamamoto/pokelingual/backend/internal/model"
)

// versionOrder defines the display order for game versions with EN/JA pairs.
var versionOrder = []string{
	"x", "y", "omega-ruby", "alpha-sapphire",
	"sun", "moon", "ultra-sun", "ultra-moon",
	"lets-go-pikachu", "lets-go-eevee",
	"sword", "shield",
}

// versionDisplayNames maps PokeAPI version names to Japanese display names.
var versionDisplayNames = map[string]string{
	"x":               "X",
	"y":               "Y",
	"omega-ruby":      "Ωルビー",
	"alpha-sapphire":  "αサファイア",
	"sun":             "サン",
	"moon":            "ムーン",
	"ultra-sun":       "Uサン",
	"ultra-moon":      "Uムーン",
	"lets-go-pikachu": "ピカブイ",
	"lets-go-eevee":   "ピカブイ",
	"sword":           "ソード",
	"shield":          "シールド",
}

// MaxPokemonID is the maximum Pokemon ID in the supported range (Gen 1-8, excluding Legends: Arceus).
// Default is 898; can be overridden from Firestore config/app at startup.
var MaxPokemonID = 898

// DefaultExcludedPokemonIDs is the default exclusion list for new users.
// Can be overridden from Firestore config/app at startup.
// Some users have phobias of certain creatures (e.g. spiders), so these Pokemon
// are excluded by default to ensure a comfortable experience.
var DefaultExcludedPokemonIDs = []int{167, 168, 595, 596, 751, 752}

// PokeAPIService implements domain.PokemonFetcher using the PokeAPI with in-memory caching.
type PokeAPIService struct {
	cache      sync.Map
	httpClient *http.Client
}

// NewPokeAPIService creates a new PokeAPIService.
func NewPokeAPIService() *PokeAPIService {
	return &PokeAPIService{
		httpClient: &http.Client{},
	}
}

type pokeAPISpeciesResponse struct {
	ID               int `json:"id"`
	Names            []struct {
		Name     string `json:"name"`
		Language struct {
			Name string `json:"name"`
		} `json:"language"`
	} `json:"names"`
	FlavorTextEntries []struct {
		FlavorText string `json:"flavor_text"`
		Language   struct {
			Name string `json:"name"`
		} `json:"language"`
		Version struct {
			Name string `json:"name"`
		} `json:"version"`
	} `json:"flavor_text_entries"`
}

type pokeAPIPokemonResponse struct {
	Sprites struct {
		FrontDefault string `json:"front_default"`
		Other        struct {
			OfficialArtwork struct {
				FrontDefault string `json:"front_default"`
			} `json:"official-artwork"`
		} `json:"other"`
	} `json:"sprites"`
}

// TotalAvailablePokemon returns the total number of Pokemon available for quests,
// accounting for user-specific exclusions.
func TotalAvailablePokemon(userExcludedIDs []int) int {
	return MaxPokemonID - len(userExcludedIDs)
}

// GetRandomPokemon returns a random Pokemon from the supported range (ID 1-MaxPokemonID).
// Per-user exclusion filtering is handled by QuestService.
func (s *PokeAPIService) GetRandomPokemon(ctx context.Context) (*model.Pokemon, error) {
	id := rand.Intn(MaxPokemonID) + 1
	return s.GetPokemonByID(ctx, id)
}

// GetPokemonByID returns a Pokemon by ID, using a cache to avoid repeated API calls.
func (s *PokeAPIService) GetPokemonByID(ctx context.Context, id int) (*model.Pokemon, error) {
	if cached, ok := s.cache.Load(id); ok {
		return cached.(*model.Pokemon), nil
	}

	pokemon, err := s.fetchFromAPI(ctx, id)
	if err != nil {
		return nil, err
	}

	s.cache.Store(id, pokemon)
	return pokemon, nil
}

func (s *PokeAPIService) fetchFromAPI(ctx context.Context, id int) (*model.Pokemon, error) {
	speciesURL := fmt.Sprintf("https://pokeapi.co/api/v2/pokemon-species/%d", id)
	speciesReq, err := http.NewRequestWithContext(ctx, "GET", speciesURL, nil)
	if err != nil {
		return nil, fmt.Errorf("creating species request: %w", err)
	}

	speciesResp, err := s.httpClient.Do(speciesReq)
	if err != nil {
		return nil, fmt.Errorf("fetching species: %w", err)
	}
	defer func() { _ = speciesResp.Body.Close() }()

	if speciesResp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("species API returned status %d", speciesResp.StatusCode)
	}

	var species pokeAPISpeciesResponse
	if err := json.NewDecoder(speciesResp.Body).Decode(&species); err != nil {
		return nil, fmt.Errorf("decoding species: %w", err)
	}

	pokemonURL := fmt.Sprintf("https://pokeapi.co/api/v2/pokemon/%d", id)
	pokemonReq, err := http.NewRequestWithContext(ctx, "GET", pokemonURL, nil)
	if err != nil {
		return nil, fmt.Errorf("creating pokemon request: %w", err)
	}

	pokemonResp, err := s.httpClient.Do(pokemonReq)
	if err != nil {
		return nil, fmt.Errorf("fetching pokemon: %w", err)
	}
	defer func() { _ = pokemonResp.Body.Close() }()

	var pokemonData pokeAPIPokemonResponse
	if err := json.NewDecoder(pokemonResp.Body).Decode(&pokemonData); err != nil {
		return nil, fmt.Errorf("decoding pokemon: %w", err)
	}

	nameEN := ""
	nameJA := ""
	for _, name := range species.Names {
		switch name.Language.Name {
		case "en":
			nameEN = name.Name
		case "ja":
			nameJA = name.Name
		}
	}

	flavorTexts := buildFlavorTextPairs(species.FlavorTextEntries)

	if len(flavorTexts) == 0 {
		return nil, fmt.Errorf("no EN/JA description pair found for pokemon %d", id)
	}

	// Use the first pair as the default description (version-matched)
	descriptionEN := flavorTexts[0].DescriptionEN
	descriptionJA := flavorTexts[0].DescriptionJA

	spriteURL := pokemonData.Sprites.Other.OfficialArtwork.FrontDefault
	if spriteURL == "" {
		spriteURL = pokemonData.Sprites.FrontDefault
	}

	return &model.Pokemon{
		ID:            id,
		NameEN:        nameEN,
		NameJA:        nameJA,
		DescriptionEN: descriptionEN,
		DescriptionJA: descriptionJA,
		SpriteURL:     spriteURL,
		FlavorTexts:   flavorTexts,
	}, nil
}

// flavorTextsByVersion collects EN and JA texts per game version.
type flavorTextsByVersion struct {
	en     string
	ja     string
	jaHrkt string
}

// buildFlavorTextPairs groups flavor_text_entries by version, creates EN/JA pairs,
// deduplicates identical text pairs, and sorts by version order.
func buildFlavorTextPairs(entries []struct {
	FlavorText string `json:"flavor_text"`
	Language   struct {
		Name string `json:"name"`
	} `json:"language"`
	Version struct {
		Name string `json:"name"`
	} `json:"version"`
}) []model.FlavorTextPair {
	// Collect texts by version
	byVersion := map[string]*flavorTextsByVersion{}
	for _, entry := range entries {
		ver := entry.Version.Name
		if _, ok := versionDisplayNames[ver]; !ok {
			continue // skip versions we don't display
		}
		if byVersion[ver] == nil {
			byVersion[ver] = &flavorTextsByVersion{}
		}
		text := cleanFlavorText(entry.FlavorText)
		switch entry.Language.Name {
		case "en":
			byVersion[ver].en = text
		case "ja":
			byVersion[ver].ja = text
		case "ja-Hrkt":
			byVersion[ver].jaHrkt = text
		}
	}

	// Build pairs: only versions with both EN and JA (or ja-Hrkt fallback)
	type versionPair struct {
		version string
		en      string
		ja      string
	}
	var pairs []versionPair
	for ver, texts := range byVersion {
		ja := texts.ja
		if ja == "" {
			ja = texts.jaHrkt
		}
		if texts.en == "" || ja == "" {
			continue
		}
		pairs = append(pairs, versionPair{version: ver, en: texts.en, ja: ja})
	}

	// Sort by predefined version order
	orderIndex := map[string]int{}
	for i, v := range versionOrder {
		orderIndex[v] = i
	}
	sort.Slice(pairs, func(i, j int) bool {
		return orderIndex[pairs[i].version] < orderIndex[pairs[j].version]
	})

	// Deduplicate: merge versions with identical EN+JA text
	var result []model.FlavorTextPair
	for _, p := range pairs {
		displayName := versionDisplayNames[p.version]
		merged := false
		for i := range result {
			if result[i].DescriptionEN == p.en && result[i].DescriptionJA == p.ja {
				// Avoid duplicate display names (e.g. two "ピカブイ")
				alreadyHas := false
				for _, name := range result[i].VersionNames {
					if name == displayName {
						alreadyHas = true
						break
					}
				}
				if !alreadyHas {
					result[i].VersionNames = append(result[i].VersionNames, displayName)
				}
				merged = true
				break
			}
		}
		if !merged {
			result = append(result, model.FlavorTextPair{
				VersionNames:  []string{displayName},
				DescriptionEN: p.en,
				DescriptionJA: p.ja,
			})
		}
	}

	return result
}

func cleanFlavorText(text string) string {
	text = strings.ReplaceAll(text, "\f", " ")
	text = strings.ReplaceAll(text, "\n", " ")
	text = strings.ReplaceAll(text, "\r", " ")

	for strings.Contains(text, "  ") {
		text = strings.ReplaceAll(text, "  ", " ")
	}

	return strings.TrimSpace(text)
}

File: backend/src/service/pokeapi-service.ts
import type { PokemonFetcher } from "../domain/interfaces.js";
import type { Pokemon, FlavorTextPair } from "../types/index.js";

const versionOrder = [
  "x", "y", "omega-ruby", "alpha-sapphire",
  "sun", "moon", "ultra-sun", "ultra-moon",
  "lets-go-pikachu", "lets-go-eevee",
  "sword", "shield",
];

const versionDisplayNames: Record<string, string> = {
  "x": "X",
  "y": "Y",
  "omega-ruby": "Ωルビー",
  "alpha-sapphire": "αサファイア",
  "sun": "サン",
  "moon": "ムーン",
  "ultra-sun": "Uサン",
  "ultra-moon": "Uムーン",
  "lets-go-pikachu": "ピカブイ",
  "lets-go-eevee": "ピカブイ",
  "sword": "ソード",
  "shield": "シールド",
};

/** 抽選対象とする最大ポケモンID。Firestore 設定があれば起動時に上書きされる。 */
export let maxPokemonID = 898;
/** デフォルトで出題から除外するポケモンIDリスト。Firestore 設定があれば起動時に上書きされる。 */
export let defaultExcludedPokemonIDs = [167, 168, 595, 596, 751, 752];

/** maxPokemonID を更新する。起動時の Firestore 設定読み込みから呼ばれる。 */
export function setMaxPokemonID(id: number) { maxPokemonID = id; }
/** defaultExcludedPokemonIDs を更新する。起動時の Firestore 設定読み込みから呼ばれる。 */
export function setDefaultExcludedPokemonIDs(ids: number[]) { defaultExcludedPokemonIDs = ids; }

interface PokeAPISpeciesResponse {
  id: number;
  is_legendary: boolean;
  is_mythical: boolean;
  names: { name: string; language: { name: string } }[];
  flavor_text_entries: {
    flavor_text: string;
    language: { name: string };
    version: { name: string };
  }[];
}

interface PokeAPIPokemonResponse {
  sprites: {
    front_default: string;
    other: { "official-artwork": { front_default: string } };
  };
  stats: { base_stat: number }[];
  types: { type: { name: string } }[];
  height: number;
  weight: number;
}

/** PokeAPI から種別情報を取得しメモリキャッシュする PokemonFetcher 実装。 */
export class PokeAPIService implements PokemonFetcher {
  private cache = new Map<number, Pokemon>();

  /** ランダムなポケモンを 1 体取得する。 */
  async getRandomPokemon(): Promise<Pokemon> {
    const id = Math.floor(Math.random() * maxPokemonID) + 1;
    return this.getPokemonByID(id);
  }

  /** ID 指定でポケモンを取得する。キャッシュ済みなら API は叩かない。 */
  async getPokemonByID(id: number): Promise<Pokemon> {
    const cached = this.cache.get(id);
    if (cached) return cached;

    const pokemon = await this.fetchFromAPI(id);
    this.cache.set(id, pokemon);
    return pokemon;
  }

  private async fetchFromAPI(id: number): Promise<Pokemon> {
    const [speciesResp, pokemonResp] = await Promise.all([
      fetch(`https://pokeapi.co/api/v2/pokemon-species/${id}`),
      fetch(`https://pokeapi.co/api/v2/pokemon/${id}`),
    ]);

    if (!speciesResp.ok) throw new Error(`species API returned status ${speciesResp.status}`);
    if (!pokemonResp.ok) throw new Error(`pokemon API returned status ${pokemonResp.status}`);

    const species: PokeAPISpeciesResponse = await speciesResp.json();
    const pokemonData: PokeAPIPokemonResponse = await pokemonResp.json();

    let nameEN = "";
    let nameJA = "";
    for (const n of species.names) {
      if (n.language.name === "en") nameEN = n.name;
      if (n.language.name === "ja") nameJA = n.name;
    }

    const flavorTexts = buildFlavorTextPairs(species.flavor_text_entries);
    if (flavorTexts.length === 0) {
      throw new Error(`no EN/JA description pair found for pokemon ${id}`);
    }

    const bst = pokemonData.stats.reduce((sum, s) => sum + s.base_stat, 0);
    const types = pokemonData.types.map((t) => t.type.name);
    const spriteURL =
      pokemonData.sprites.other["official-artwork"].front_default ||
      pokemonData.sprites.front_default;

    return {
      id,
      name_en: nameEN,
      name_ja: nameJA,
      description_en: flavorTexts[0].description_en,
      description_ja: flavorTexts[0].description_ja,
      sprite_url: spriteURL,
      base_stat_total: bst,
      types,
      height: pokemonData.height,
      weight: pokemonData.weight,
      is_legendary: species.is_legendary,
      is_mythical: species.is_mythical,
      flavor_texts: flavorTexts,
    };
  }
}

interface FlavorTextsByVersion {
  en: string;
  ja: string;
  jaHrkt: string;
}

function buildFlavorTextPairs(
  entries: PokeAPISpeciesResponse["flavor_text_entries"],
): FlavorTextPair[] {
  const byVersion = new Map<string, FlavorTextsByVersion>();

  for (const entry of entries) {
    const ver = entry.version.name;
    if (!(ver in versionDisplayNames)) continue;

    if (!byVersion.has(ver)) {
      byVersion.set(ver, { en: "", ja: "", jaHrkt: "" });
    }
    const texts = byVersion.get(ver)!;
    const cleaned = cleanFlavorText(entry.flavor_text);

    switch (entry.language.name) {
      case "en": texts.en = cleaned; break;
      case "ja": texts.ja = cleaned; break;
      case "ja-Hrkt": texts.jaHrkt = cleaned; break;
    }
  }

  interface VersionPair { version: string; en: string; ja: string }
  const pairs: VersionPair[] = [];

  for (const [ver, texts] of byVersion) {
    const ja = texts.ja || texts.jaHrkt;
    if (!texts.en || !ja) continue;
    pairs.push({ version: ver, en: texts.en, ja });
  }

  const orderIndex = new Map(versionOrder.map((v, i) => [v, i]));
  pairs.sort((a, b) => (orderIndex.get(a.version) ?? 999) - (orderIndex.get(b.version) ?? 999));

  const result: FlavorTextPair[] = [];
  for (const p of pairs) {
    const displayName = versionDisplayNames[p.version];
    const existing = result.find(
      (r) => r.description_en === p.en && r.description_ja === p.ja,
    );
    if (existing) {
      if (!existing.version_names.includes(displayName)) {
        existing.version_names.push(displayName);
      }
    } else {
      result.push({
        version_names: [displayName],
        description_en: p.en,
        description_ja: p.ja,
      });
    }
  }

  return result;
}

function cleanFlavorText(text: string): string {
  return text
    .replace(/\f/g, " ")
    .replace(/\n/g, " ")
    .replace(/\r/g, " ")
    .replace(/ {2,}/g, " ")
    .trim();
}
